src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -131,10 +131,25 @@ private async Task InvokeInternal(HttpContext context)
         ResponseBufferingStream? responseBufferingStream = null;
         IHttpResponseBodyFeature? originalBodyFeature = null;
 
+        UpgradeFeatureLoggingDecorator? loggableUpgradeFeature = null;
+        IHttpUpgradeFeature? originalUpgradeFeature = null;
+
         try
         {
             var response = context.Response;
 
+            if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponsePropertiesAndHeaders))
+            {
+                originalUpgradeFeature = context.Features.Get<IHttpUpgradeFeature>();
+
+                if (originalUpgradeFeature != null && originalUpgradeFeature.IsUpgradableRequest)
+                {
+                    loggableUpgradeFeature = new UpgradeFeatureLoggingDecorator(originalUpgradeFeature, () => LogResponseHeaders(context.Response, options, _logger));
```

Comment:
Don't pass in a delegate, follow the pattern used by the body logging:
https://github.com/dotnet/aspnetcore/blob/79a19b7f74d1f8630462b07b4e803e667a618fe7/src/Middleware/HttpLogging/src/ResponseBufferingStream.cs#L128

Reply:
I used this design in order to be able to test the `UpgradeFeatureLoggingDecorator` class. If I call directly that static method, it becomes harder to test. Is it acceptable? Should I remove the tests of that class altogether since there is no perceived value?

Reply:
That test class isn't adding much value, remove it.